src/smt/theory_datatype.h

```diff
@@ -73,11 +72,58 @@ namespace smt {
         void propagate_recognizer(theory_var v, enode * r);
         void sign_recognizer_conflict(enode * c, enode * r);
 
-        ptr_vector<enode>    m_to_unmark;
-        enode_pair_vector    m_used_eqs;
-        enode *              m_main;
+        typedef enum { ENTER, EXIT } stack_op;
+        typedef map<enode*, enode*, obj_ptr_hash<enode>, ptr_eq<enode> > parent_tbl;
```

Comment:
I have to confess that map<enode* made me worried, but I can see it is fine given that you use obj_ptr_hash.
Just don't use ptr_hash as it leads to non-determinism across platforms and runs if there is any code that depends on iterating over the table.